design-proposals/golden-image-delivery-and-update-pipeline.md

```diff
@@ -0,0 +1,207 @@
+
+A **VolumeReference** object references an import/clone source for a PVC. Functionally, it is similar to the “source” field of a DataVolume. The VolumeReference’s source is mutable, which has the side effect of allowing the VolumeReference to act as a static link that will redirect to whatever source the link currently points to. In the future, the VolumeReference will act as a PVC Populator as well as a DataVolume source.
+
+**VolumeImportPollers can manage a corresponding VolumeReference** to always point to the latest most up-to-date PVC that the poller has created. 
+
+
+## CDI: DataVolume VolumeReference Source Field
+
+The DataVolume object will add a new source field capable of importing pvc content using a VolumeReference. 
+
+![DataVolume VolumeReference Source Field](golden-image-delivery-images/volume-reference-source.png "DataVolume VolumeReference Source Field")
+
+## SSP: Setup VolumeImportPollers for common templates
+
+Since the SSP operator manages installing common templates, this operator should also be responsible for setting up the VolumeImportPollers used to automatically source the disk images into the cluster that the common templates use.
+
+These VolumeImportPollers can be configured on the SSP operators Custom Resource.
+
+
+![SSP Disk Import Pipelines](golden-image-delivery-images/ssp-disk-import-pipelines.png "SSP Disk Import Pipelines")
+
+
+## Common templates: VolumeReferences as DataVolume Source
+
+The common templates shipped with SSP need to have their DataVolumeTemplate sections transition to using a VolumeReference as the source field rather than a PVC claim name. This allows the VolumeImportPoller to seamlessly update the underlying PVC the corresponding VolumeReference points to, which then ensures that the common templates are always using the most up-to-date disk images.
+
+## HCO: Integration with SSP Operator
+
+The HCO is responsible for setting VolumeImportPoller sources on the SSP operator’s CR. These volume sources are the locations used by the VolumeImportPollers to populate PVCs for the common templates.
+
+
+## OCP UI: Integration with VolumeReferences
+
+The UI currently detects if a template has a boot source by checking if the DataVolumeTemplate’s source PVC is present. The UI needs to now be capable of checking if a DataVolumeTemplate’s source VolumeReference has phase: linked as well in order to determine if a boot source is set.
+
+## Scalability
+
+The only scaling concern is that the automated process has the potential to exhaust storage resources unless a PVC garbage collection mechanism is observed by the VolumeImportPoller.
+
+## Update/Rollback Compatibility
+
+* SSP operator must make image pipelines an opt-in feature to ensure existing boot images are not overwritten. 
```

Comment:
The VolumeImportPoller could have a pause field or only import once if the interval is set to null.
This way SSP and the templates could create them but the fetching would only happen once. When the user enables the autoupdate, it would just set an interval.
On the other hand this would not result in a that stable image in cases where the operator CR gets reinstalled.

Can the operator also just create the VolumeReferences and if the admin enables auto-update we just create the poller pointing to the respective ref?

Reply:
> Can the operator also just create the VolumeReferences and if the admin enables auto-update we just create the poller pointing to the respective ref?

I think the most straightforward approach is for SSP to create VolumeReferences by default that match whatever is used by the common templates. SSP will simply always do this no matter what.

When an admin opts-in to using auto delivery/update, SSP would be configured to create a VolumeImportPoller which can then adopt ownership of the corresponding VolumeReference. If the admin later opts-out of the feature, SSP can delete the VolumeImportPoller and orphan the VolumeReference (use the equivalent of cascade=false for deletion) which will leave the VolumeReference intact. 

I'll update the doc to talk about this.